File: Swift/ExampleTests/common/AddressString.swift
public struct AddressString: CustomStringConvertible, Equatable {
    public let street: String
    public let city: String
    public let state: String
    public let zIP: String

    public init(street: String, city: String, state: String, zIP: String) {
        self.street = street
        self.city = city
        self.state = state
        self.zIP = zIP
    }

    public init(fromArray v: [String]) {
        self.street = v.count > 0 ? v[0] : ""
        self.city = v.count > 1 ? v[1] : ""
        self.state = v.count > 2 ? v[2] : ""
        self.zIP = v.count > 3 ? v[3] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> AddressString {
        let parts = Tokens.require(text, 4, "Address")
        return AddressString(street: parts[0], city: parts[1], state: parts[2], zIP: parts[3])
    }

    public var description: String {
        return Tokens.token(street) + " " + Tokens.token(city) + " " + Tokens.token(state) + " " + Tokens.token(zIP)
    }

    public static let dncString = "?DNC?"

    public static func == (a: AddressString, b: AddressString) -> Bool {
        return (a.street == dncString || b.street == dncString || a.street == b.street)
            && (a.city == dncString || b.city == dncString || a.city == b.city)
            && (a.state == dncString || b.state == dncString || a.state == b.state)
            && (a.zIP == dncString || b.zIP == dncString || a.zIP == b.zIP)
    }
}

File: Python/ExampleTests/common/address_string.py
from . import tokens
DNC_STRING = '?DNC?'


class AddressString:
    def __init__(self, street: str = '', city: str = '', state: str = '', zip: str = ''):
        self.street = street
        self.city = city
        self.state = state
        self.zip = zip

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else '',
            v[3] if len(v) > 3 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 4, 'Address')
        return cls(
            parts[0],
            parts[1],
            parts[2],
            parts[3]
        )

    def __str__(self):
        return (tokens.token(self.street) + ' ' +
                tokens.token(self.city) + ' ' +
                tokens.token(self.state) + ' ' +
                tokens.token(self.zip))

    def _key(self):
        return (self.street, self.city, self.state, self.zip)

    def __eq__(self, other):
        if not isinstance(other, AddressString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'AddressString({self})'
